test/src/test/java/org/corfudb/integration/LogReplicationIT.java

```diff
@@ -1046,6 +1051,145 @@ private void testSnapshotSyncAndLogEntrySync(int numCyclesToDelayApply, boolean
+    * immediately, the Source gets STOP event followed by log_entry Sync event. This tests that even
+    * though the metadata on Source and destination do not match, the sink is able to process the ms
+    *
+    **/
+    @Test
+    public void testLogEntrySyncWithFSMChangeAndWithAckDrop() throws Exception {
+        setupEnv();
+        // Write data in transaction to t0 and t1
+        Set<String> crossTables = new HashSet<>();
+        crossTables.add(t0);
+        crossTables.add(t1);
+
+        // Writes transactions to t0, t1 and t2 + transactions across 'crossTables'
+        writeCrossTableTransactions(crossTables, true);
+
+        // Start Log Entry Sync
+        expectedAckMessages = NUM_KEYS * WRITE_CYCLES;
+        testConfig.clear().setDropAckLevel(2);
+
+        Set<WAIT> waitCondition = new HashSet<>();
+        waitCondition.add(WAIT.NONE);
+        startLogEntrySync(crossTables, waitCondition, true, () -> changeState());
+
+        blockUntilFSMTransition.await();
+
+        checkStateChange(logReplicationSourceManager.getLogReplicationFSM(), LogReplicationStateType.INITIALIZED, true);
+        testConfig.clear();
+
+        // add a listner to ACKs received. This is used to unblock the current thread before the final verification.
+        ackMessages = sourceDataSender.getAckMessages();
+        ackMessages.addObserver(this);
+
+        // Use the last received ACK to add a new event in FSM queue.
+        LogReplicationEntryMsg ack = sourceDataSender.getAckMessages().getDataMessage();
+        logReplicationSourceManager.getLogReplicationFSM().input(new LogReplicationEvent(LogReplicationEvent.LogReplicationEventType.LOG_ENTRY_SYNC_REQUEST,
+                new LogReplicationEventMetadata(getUUID(ack.getMetadata().getSyncRequestId()), ack.getMetadata().getTimestamp(), ack.getMetadata().getSnapshotTimestamp())));
+        checkStateChange(logReplicationSourceManager.getLogReplicationFSM(), LogReplicationStateType.IN_LOG_ENTRY_SYNC, true);
+
+        // To ensure a clear start, wait until the pendingMsg queue is reset.
+        while(logReplicationSourceManager.getLogReplicationFSM().getLogEntrySender().getPendingACKQueueSize() != 0) {
+        }
+        sourceDataSender.resetTestConfig(testConfig);
+
+        expectedAckTimestamp = srcDataRuntime.getAddressSpaceView().getLogTail();
+
+        // Block until the expected ACK Timestamp is reached
+        try {
+            blockUntilExpectedAckTs.acquire();
+        } finally {
+            blockUntilExpectedAckTs.release();
+        }
+
+        // Verify Data on Destination site
+        log.debug("****** Verify Data on Destination");
+        // Because t2 is not specified as a replicated table, we should not see it on the destination
+        srcDataForVerification.get(t2).clear();
+
+        // Verify Destination
+        verifyData(dstCorfuTables, srcDataForVerification);
+    }
+
+
+    @Test
+    public void testSnapshotSyncWithAckDrops() throws Exception {
```

Comment:
Snapshot msgs are sent through pendingQueue (same as for delta sync), and msgs are removed from pendingQueue after an ACK is received for a msg or a higher ACK is received in which case msg.timestamp <= ACK.timestamp are removed from the queue.
But ACK drops for snapshot is not concerning because our snapshot does not change, which translates to the same msg being resent to Sink which can be ignored if the Sink has already processed it. Since it's still a valid scenario, I included this simple test.